File: finlib-cpp/src/finlib/mortgage/mortgage.cpp
//
// Created by Andy Pack on 19/06/2025.
//

#include "mortgage.hpp"

namespace finlib {
    Mortgage::Mortgage(double purchase_price,
                       double deposit,
                       double interest_rate,
                       int32_t term_years) {
        handle = finlibrs::mortgage_new(purchase_price, deposit, interest_rate, term_years);
    }

    double Mortgage::ltv() const {
        return finlibrs::mortgage_ltv(handle);
    }

    double Mortgage::loan_value() const {
        return finlibrs::mortgage_loan_value(handle);
    }

    double Mortgage::monthly_payment() const {
        return finlibrs::mortgage_monthly_payment(handle);
    }

    double Mortgage::total_repayment() const {
        return finlibrs::mortgage_total_repayment(handle);
    }

    double Mortgage::total_interest_repayment() const {
        return finlibrs::mortgage_total_interest_repayment(handle);
    }

    double Mortgage::present_value() const {
        return finlibrs::mortgage_present_value(handle);
    }

    double Mortgage::future_value(double annual_interest_rate) const {
        return finlibrs::mortgage_future_value(handle, annual_interest_rate);
    }

    double Mortgage::net_future_value_interest(double annual_interest_rate) const {
        return finlibrs::mortgage_net_future_value_interest(handle, annual_interest_rate);
    }

    double Mortgage::total_interest(double annual_interest_rate) const {
        return finlibrs::mortgage_total_interest(handle, annual_interest_rate);
    }

    Mortgage::~Mortgage() {
        finlibrs::mortgage_destroy(handle);
    }
} // finlib

File: FinLib.NET/FinLib/FixedIncome/Mortgage.cs
using System;

namespace FinLib.FixedIncome;

public class Mortgage: IDisposable
{
    private readonly unsafe Mortgage_native* _handle;
    internal unsafe Mortgage_native* GetPtr() => _handle;

    public Mortgage(double purchasePrice, double deposit, double interestRate, int termYears)
    {
        unsafe
        {
            _handle = NativeMethods.mortgage_new(purchasePrice, deposit, interestRate, termYears);
        }
    }

    public double LTV
    {
        get
        {
            unsafe
            {
                return NativeMethods.mortgage_ltv(_handle);
            }
        }
    }

    public double LoanValue
    {
        get
        {
            unsafe
            {
                return NativeMethods.mortgage_loan_value(_handle);
            }
        }
    }

    public double MonthlyPayment
    {
        get
        {
            unsafe
            {
                return NativeMethods.mortgage_monthly_payment(_handle);
            }
        }
    }

    public double TotalRepayment
    {
        get
        {
            unsafe
            {
                return NativeMethods.mortgage_total_repayment(_handle);
            }
        }
    }

    public double TotalInterestRepayment
    {
        get
        {
            unsafe
            {
                return NativeMethods.mortgage_total_interest_repayment(_handle);
            }
        }
    }

    public double PresentValue
    {
        get
        {
            unsafe
            {
                return NativeMethods.mortgage_present_value(_handle);
            }
        }
    }

    public double FutureValue(double annualInterestRate)
    {
        unsafe
        {
            return NativeMethods.mortgage_future_value(_handle, annualInterestRate);
        }
    }

    public double NetFutureValueInterest(double annualInterestRate)
    {
        unsafe
        {
            return NativeMethods.mortgage_net_future_value_interest(_handle, annualInterestRate);
        }
    }

    public double TotalInterest(double annualInterestRate)
    {
        unsafe
        {
            return NativeMethods.mortgage_total_interest(_handle, annualInterestRate);
        }
    }

    private void ReleaseUnmanagedResources()
    {
        unsafe
        {
            NativeMethods.mortgage_destroy(_handle);
        }
    }

    public void Dispose()
    {
        ReleaseUnmanagedResources();
        GC.SuppressFinalize(this);
    }

    ~Mortgage()
    {
        ReleaseUnmanagedResources();
    }
}